ludwig/schema/config_object.py

```diff
@@ -0,0 +1,501 @@
+        this, any global defaults that the user specified will be set here accordingly, however if no global
+        defaults were specified, each section already reflects the regular defaults for each feature, so the output
+        features will initialize as expected. Any values set on the output feature itself will be set later in
+        _set_output_features().
+
+        Args:
+            feature_dicts: List of output feature definitions in user defined config dict.
+
+        Returns:
+            None -> Updates ModelConfig.
+        """
+        for feature_dict in feature_dicts:
+            # Retrieve output feature schema cls from registry to init feature
+            feature_config = copy.deepcopy(get_output_feature_cls(feature_dict[TYPE])())
+
+            # Set global defaults on output feature config cls - if user has defined global defaults, these
+            # will be set on the feature_config class, otherwise the global defaults already reflect the regular
+            # defaults, so it will initialize the feature as expected.
+            type_defaults = getattr(
+                self.defaults, feature_dict[TYPE]
+            )  # Global defaults section for specific feature type
+            feature_config.decoder = copy.deepcopy(type_defaults.decoder)
+            feature_config.loss = copy.deepcopy(type_defaults.loss)
+
+            # TODO depending on defaults section conversation may want to enable
+            # feature_config.preprocessing = copy.deepcopy(type_defaults.preprocessing)
+
+            # Assign feature on output features container
+            setattr(self.output_features, feature_dict[NAME], feature_config)
+
+    def _set_output_features(self, feature_dicts: List[dict]) -> None:
+        """This function sets the values on the ModelConfig that are specified on the output features themselves.
+        This will override any global defaults that have been set in the previous function call
+        _initialize_output_features().
+
+        Args:
+            feature_dicts: List of output feature definitions in user defined config dict.
+
+        Returns:
+            None -> Updates ModelConfig.
+        """
+        for feature_dict in feature_dicts:
+            # Set the parameters that the user specified on the output feature itself
+            self._set_attributes(
+                getattr(getattr(self, OUTPUT_FEATURES), feature_dict[NAME]),
+                feature_dict,
+                feature_type=feature_dict[TYPE],
+            )
+
+            # Set the reduce_input parameter for the tagger decoder specifically
+            if getattr(getattr(self, OUTPUT_FEATURES), feature_dict[NAME]).decoder.type == "tagger":
+                getattr(getattr(self, OUTPUT_FEATURES), feature_dict[NAME]).reduce_input = None
+
+    def _set_attributes(self, config_obj_lvl: BaseMarshmallowConfig, config_dict_lvl: dict, feature_type: str = None):
```

Comment:
Maybe consider factoring this out into a stateless utility in config_object_utils with simple tests to demonstrate what this is doing.

It looks like it's doing something cool and straightforward that I would understand from a glance at a unit test, while the function itself is rather complex.

Reply:
In this instance, if we were to do this, I would probably want to refactor all the other util functions and port them over to a util file as well (i.e. initialize_input_features, set_input_features, output_features, get_config_cls, etc.) This would be a considerable amount of effort and would change the structure of the config object a bit. Since this function is kind of the core utility of the config object, it is tested pretty directly under the `test_config_object` test and indirectly  through the rest of the tests. Would you be open to taking 10-15 with me to do a walk through instead? Additionally, once this lands, I plan on either taking a tech talk slot or an all hands slot to walk the entire team through the config object flow so that everyone understands what's going on. How does this sound?